internal/state/graph.go

```diff
@@ -88,7 +123,91 @@ func buildGraph(
 		}
 	}
 
-	return g
+	return g, warnings
+}
+
+func resolveBackends(
+	ctx context.Context,
+	routes map[types.NamespacedName]*route,
+	services map[types.NamespacedName]*v1.Service,
+	serviceResolver resolver.ServiceResolver,
+) (map[backendService]backend, Warnings) {
+	warnings := newWarnings()
+	backends := make(map[backendService]backend)
+
+	for _, r := range routes {
+
+		backendServicesForRoute := make(map[ruleIndex]backendService)
+
+		for i, rule := range r.Source.Spec.Rules {
+
+			backendSvc, err := getBackendServiceFromRouteRule(rule, r.Source.Namespace)
+			backendServicesForRoute[ruleIndex(i)] = backendSvc
+
+			if err != nil {
+				warnings.AddWarningf(r.Source, "invalid backend ref for rule %d: %s", i, err.Error())
+
+				continue
+			}
+
+			if b, exists := backends[backendSvc]; exists {
+				if b.ErrorMsg != "" {
+					warnings.AddWarningf(r.Source, "cannot resolve backend ref for rule %d: %s", i, b.ErrorMsg)
+				}
+
+				continue
+			}
+
+			b := backend{}
+
+			svcName := types.NamespacedName{Namespace: backendSvc.Namespace, Name: backendSvc.Name}
+
+			svc, exists := services[svcName]
+
+			if exists {
+				b.Endpoints, err = serviceResolver.Resolve(ctx, svc, backendSvc.Port)
+				if err != nil {
+					b.ErrorMsg = err.Error()
+					warnings.AddWarningf(r.Source, "cannot resolve backend ref for rule %d: %s", i, err.Error())
```

Comment:
We print out "cannot resolve backend ref for rule..." in three (maybe more?) different branches.

The 2 here are from Resolve and the one above due to a missing backend. Is Error enough to discern what's happening? Do missing backends and an error from Resolve ulitmately have different root causes?

Reply:
I collapsed the two resolve warnings, but I think the remaining warnings are necessary.

The warnings are a bit repetitive, but I'm trying to prevent making duplicate API calls to the k8s API server. Since rules and routes can reference the same backend service, I'm checking if the backend service has already been resolved before attempting to resolve it. 

I also think an invalid backend ref is different from a resolve failure. An invalid backend ref means the HTTPRoute config is invalid, whereas a resolve error could be a network error or a Service configuration error.

Reply:
>  I'm trying to prevent making duplicate API calls to the k8s API server.

just to double check. I don't think we make any API calls -- the Mgr client is a caching client and it should return EndpointSlices from its cache (shared informer for EndpointSlices). Do you see any cases when the API calls will still happen?  If that is case, I think we have a problem because an API call is an expensive operation.

Reply:
You're right; it is a caching client.